bin/Macronify/Main.cpp

```diff
@@ -0,0 +1,734 @@
+                }
+            }
+            return op;
+        }
+    };
+
+    template< typename Derived >
+    using VisitorConfig = vast::cg::CodeGenFallBackVisitorMixin< Derived,
+        CodeGenVisitorMixin,
+        vast::cg::DefaultFallBackVisitorMixin
+    >;
+
+    using Visitor =
+        vast::cg::CodeGenVisitor<VisitorConfig, MacroniMetaGenerator>;
+
+} // namespace macroni
+
+static llvm::cl::list< std::string > compiler_args(
+    "ccopts", llvm::cl::ZeroOrMore, llvm::cl::desc("Specify compiler options")
+);
+
+int main(int argc, char **argv) {
+
+    pasta::InitPasta initializer;
+    pasta::FileManager fm(pasta::FileSystem::CreateNative());
+    auto maybe_compiler =
+        pasta::Compiler::CreateHostCompiler(fm, pasta::TargetLanguage::kCXX);
+    if (!maybe_compiler.Succeeded()) {
+        std::cerr << maybe_compiler.TakeError() << std::endl;
+        return EXIT_FAILURE;
+    }
+
+    auto maybe_cwd = (pasta::FileSystem::From(maybe_compiler.Value())
+                      ->CurrentWorkingDirectory());
+    if (!maybe_cwd.Succeeded()) {
+        std::cerr << maybe_compiler.TakeError() << std::endl;
+        return EXIT_FAILURE;
+    }
+
+    const pasta::ArgumentVector args(argc - 1, &argv[1]);
+    auto maybe_command = pasta::CompileCommand::CreateFromArguments(
+        args, maybe_cwd.TakeValue());
+    if (!maybe_command.Succeeded()) {
+        std::cerr << maybe_command.TakeError() << std::endl;
+        return EXIT_FAILURE;
+    }
+
+    const auto command = maybe_command.TakeValue();
+    auto maybe_jobs = maybe_compiler->CreateJobsForCommand(command);
+    if (!maybe_jobs.Succeeded()) {
+        std::cerr << maybe_jobs.TakeError() << std::endl;
+        return EXIT_FAILURE;
+    }
+
+    for (const auto &job : maybe_jobs.TakeValue()) {
+        auto maybe_ast = job.Run();
+        if (!maybe_ast.Succeeded()) {
+            std::cerr << maybe_ast.TakeError() << std::endl;
+            return EXIT_FAILURE;
+        }
+
+        ast.emplace(maybe_ast.TakeValue());
+
+        mlir::DialectRegistry registry;
+
+        // Register the MLIR dialects we will be converting to
+        registry.insert<
+            vast::hl::HighLevelDialect,
+            macroni::macroni::MacroniDialect,
+            macroni::kernel::KernelDialect
+        >();
+        mctx.emplace(registry);
+        macroni::MacroniMetaGenerator meta(*ast, &*mctx);
+        vast::cg::CodeGenBase<macroni::Visitor> codegen(&*mctx, meta);
+        builder.emplace(&*mctx);
+
+        codegen.append_to_module(ast->UnderlyingAST().getTranslationUnitDecl());
+        mlir::OwningOpRef<mlir::ModuleOp> mod = codegen.freeze();
+
+        // TODO(bpp): Add a command-line argument to convert special macros into
+        // special operations
+
+        // Register conversions
+
+        // Mark expansions of get_user(), offsetof() and container_of() as
+        // illegal
+        mlir::ConversionTarget trg(*mctx);
```

Comment:
Each of these should be a pulled out into separate fn/lambda, as it makes it much easier to read.